src/main/java/com/solvd/micro9/tickets/service/impl/EsTicketCommandHandlerImpl.java

```diff
@@ -0,0 +1,98 @@
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.util.function.Tuple2;
+
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+@Service
+@Slf4j
+@RequiredArgsConstructor
+public class EsTicketCommandHandlerImpl implements EsTicketCommandHandler {
+
+    public static final String TICKET_USER_ID_TO_NULL_PAYLOAD = "{\"userId\":null}";
+    private final EsTicketRepository esTicketRepository;
+    private final EsEventRepository esEventRepository;
+    private final KfProducer producer;
+
+    @Transactional
+    @Override
+    public Mono<EsTicket> apply(CreateTicketCommand command) {
+        String payload = new Gson().toJson(command.getTicket());
+        EsTicket event = EsTicket.builder()
+                .type(EsEventType.TICKET_CREATED)
+                .time(LocalDateTime.now())
+                .createdBy(command.getCommandBy())
+                .entityId(UUID.randomUUID().toString())
+                .payload(payload)
+                .build();
+
+        return esEventRepository.findByEntityId(command.getTicket().getEventId())
+                .collectList()
+                .map(esEventsList -> {
+                    if (esEventsList.isEmpty()) {
+                        throw new ResourceDoesNotExistException(
+                                "Event [id=" + command.getTicket().getEventId() + "] does not exist"
+                        );
+                    } else {
+                        return esEventsList;
+                    }
+                })
+                .zipWith(esTicketRepository.save(event))
+                .map(Tuple2::getT2)
+                .doOnSuccess(esTicket -> producer.send("New event", esTicket));
+    }
+
+    @Override
+    public void apply(SetTicketsUserIdToNullByUserIdCommand command) {
+        List<EsTicket> esTicketList = new ArrayList<>();
+        final boolean[] isStreamCompleted = {false};
+
+        esTicketRepository.findAll()
+                .filter(esTicket -> {
+                    Ticket ticket = new Gson().fromJson(esTicket.getPayload(), Ticket.class);
+                    return command.getUserId().equals(ticket.getUserId());
+                })
+                .doOnNext(esTicket -> {
+                    EsTicket event = EsTicket.builder()
+                            .type(EsEventType.TICKET_UPDATED)
+                            .time(LocalDateTime.now())
+                            .createdBy(command.getCommandBy())
+                            .entityId(esTicket.getEntityId())
+                            .payload(TICKET_USER_ID_TO_NULL_PAYLOAD)
+                            .build();
+                    esTicketList.add(event);
+                })
+                .doOnComplete(() -> isStreamCompleted[0] = true)
+                .subscribe();
+
+        while (!isStreamCompleted[0]) {
+        } //TODO is there a better way to do?
```

Comment:
@eyakauleva I will pull it and report to you

Reply:
@eyakauleva try to use `flatMap` or `handle`?
https://projectreactor.io/docs/core/release/reference/#_handle

```java
esTicketRepository.findAll()
.handle() // filtering and mapping
.subscribeOn(Schedulers.boundElastic())
                                 ...

```

Reply:
@h1alexbel flatMap not working

just tried with handle and that's not working neither

option 1:
```java
public void apply(SetTicketsUserIdToNullByUserIdCommand command) {
        Flux<EsTicket> esTicketFlux = esTicketRepository.findAll()
                .handle((esTicket, sink) -> {
                    Ticket ticket = new Gson().fromJson(esTicket.getPayload(), Ticket.class);
                    EsTicket event = EsTicket.builder()
                            .type(EsType.TICKET_USER_DELETED)
                            .time(LocalDateTime.now())
                            .createdBy(command.getCommandBy())
                            .entityId(esTicket.getEntityId())
                            .build();
                    if(esTicket.getType().equals(EsType.TICKET_CREATED)
                            && command.getUserId().equals(ticket.getUserId())) {
                        sink.next(event);
                    }
                });
        esTicketFlux.subscribe(esTicket -> {
            esTicketRepository.save(esTicket).subscribe();
        });
    }
```
option 2:
```java
public void apply(SetTicketsUserIdToNullByUserIdCommand command) {
        esTicketRepository.findAll()
                .handle((esTicket, sink) -> {
                    Ticket ticket = new Gson().fromJson(esTicket.getPayload(), Ticket.class);
                    EsTicket event = EsTicket.builder()
                            .type(EsType.TICKET_USER_DELETED)
                            .time(LocalDateTime.now())
                            .createdBy(command.getCommandBy())
                            .entityId(esTicket.getEntityId())
                            .build();
                    if(esTicket.getType().equals(EsType.TICKET_CREATED)
                            && command.getUserId().equals(ticket.getUserId())) {
                       esTicketRepository.save(event).subscribe();
                    }
                })
                .subscribe();
    }
```

Reply:
@eyakauleva forked, I'll create an issue